### ISSUE TITLE
Process @ params on cli command line by reading file/URL

### ISSUE THREAD

**Issue by mandarjog** [cloudfoundry/cli#734]
Enable @ params to enable command like the following
```
cf  set-env spring-music BLAH @abc.json
```

Where abc.json is
```
{
    "aa": 200,
    "bb": "xyz"
}
```

It also makes other command easy to use such as 
```
cf create-user-provided-service my-db-mine -p '{"username":"admin","password":"pa55woRD"}'
```
to
```
cf create-user-provided-service my-db-mine -p @crentials.json 
```

**Comment by dkoper** [comment]
Hi @mandarjog 

Thanks for bringing this up.
We have a few stories around enabling a user to pass in input from a file instead of inline, the other way around, or either, summarized in #650. I have added set-env to the list.

We like consistency, and can't introduce changes in a CLI v6 release that could break existing scripts.
The '@' approach is consistent with the preferred solution for `cf curl` because of curl's use of `-d @file`.
However, it would break compatibility for users already scripting things like `cf set-env myapp domain-separator @`.
I will look into this further.

**Comment by mandarjog** [comment]
The implementation only looks for patan value starting with a '@'
So domains will be fine.

It also ensures that if a file with that name is not available, the original value is retained. So existing scripts should not break.

Regards, 
Sent from my Ericsson Blaze

Mandar U Jog

**Comment by dkoper** [comment]
How about `cf login -u user -p @abc123` - makes the CLI access the file system?
I don't think it's a good idea to introduce this a feature to *all* parameters, it's easy to overlook its impact - I'd prefer to introduce it to commands that users actually need it for.

**Comment by mandarjog** [comment]
I understand from the backward compatibility standpoint it makes sense.

However the description
"If a param is prefixed with "@",  Cli treats it as a file indirection"

1. Is far more succinct.
2. It is also automatically applied to all cli / plugins. 
3. It is also *very* consistent way of dealing with indirections.
4. Leaves the door open for general URL indirection.

I understand the issue with unanticipated change in behavior, can that be addressed with a global flag like "param-indirection=false" ?


Regards, 
Sent from my Ericsson Blaze

Mandar U Jog

**Comment by dkoper** [comment]
Few people know about/look for global flags. They're usually to configure the http proxy to use, set timeouts or debugging statements.

Users have only expressed a strong need for support for reading files for curl, cups/uups and now set-env.

I think we should go forward with the implementation of the feature, but only for these commands.

In v7, where we can introduce breaking changes, we can consider a new, consistent approach, and deprecate the current strategy used in create-service/update-service/bind-service/create-service-key so that we don't have two ways of doing the same thing.

Cheers,
Dies Koper
CF CLI PM

**Comment by mandarjog** [comment]
Ok great.
Thanks for the clarification.

Regards, 
Sent from my Ericsson Blaze

Mandar U Jog

**Comment by krishicks** [comment]
We won't be pulling this in; it is too broad a change and would introduce confusion with commands that presently take a file parameter.

Additionally, I'm not keen on the implementation: a more appropriate thing for it to do would be to provide a Reader that could be used by a command. Reading the entire file at once isn't what we would necessarily want to do.

**Comment by krishicks** [comment]
I've closed #735. I don't think this @file pattern is very widespread. I've seen only `curl` support it, and even then only when it's given as the value to the `--data` flag, not as an arg.
